beacon_node/network/src/sync/sampling.rs

```diff
@@ -0,0 +1,538 @@
+        // request was sent, loop to increase the required_successes until the sampling fails if
+        // there are no peers.
+        if ongoings == 0 && sent_requests == 0 {
+            debug!(self.log, "Sampling request stalled"; "block_root" => %self.block_root);
+        }
+
+        Ok(None)
+    }
+}
+
+mod request {
+    use super::{SamplingError, SamplingId, SamplingRequester};
+    use crate::sync::{
+        manager::DataColumnsByRootRequester,
+        network_context::{DataColumnsByRootSingleBlockRequest, SyncNetworkContext},
+    };
+    use beacon_chain::BeaconChainTypes;
+    use lighthouse_network::PeerId;
+    use std::collections::HashSet;
+    use types::{data_column_sidecar::ColumnIndex, EthSpec, Hash256, Slot};
+
+    pub(crate) struct ActiveColumnSampleRequest {
+        column_index: ColumnIndex,
+        status: Status,
+        // TODO(das): Should downscore peers that claim to not have the sample?
+        #[allow(dead_code)]
+        peers_dont_have: HashSet<PeerId>,
+    }
+
+    #[derive(Debug, Clone)]
+    enum Status {
+        NoPeers,
+        NotStarted,
+        Sampling(PeerId),
+        Verified,
+    }
+
+    impl ActiveColumnSampleRequest {
+        pub(crate) fn new(column_index: ColumnIndex) -> Self {
+            Self {
+                column_index,
+                status: Status::NotStarted,
+                peers_dont_have: <_>::default(),
+            }
+        }
+
+        pub(crate) fn is_completed(&self) -> bool {
+            match self.status {
+                Status::NoPeers | Status::NotStarted | Status::Sampling(_) => false,
+                Status::Verified => true,
+            }
+        }
+
+        pub(crate) fn is_failed(&self) -> bool {
+            match self.status {
+                Status::NotStarted | Status::Sampling(_) | Status::Verified => false,
+                Status::NoPeers => true,
+            }
+        }
+
+        pub(crate) fn is_ongoing(&self) -> bool {
+            match self.status {
+                Status::NotStarted | Status::NoPeers | Status::Verified => false,
+                Status::Sampling(_) => true,
+            }
+        }
+
+        pub(crate) fn request<T: BeaconChainTypes>(
+            &mut self,
+            block_root: Hash256,
+            block_slot: Slot,
+            requester: SamplingRequester,
+            cx: &mut SyncNetworkContext<T>,
+        ) -> Result<bool, SamplingError> {
+            match &self.status {
+                Status::NoPeers | Status::NotStarted => {} // Ok to continue
+                Status::Sampling(_) => return Ok(false),   // Already downloading
+                Status::Verified => return Ok(false),      // Already completed
+            }
+
+            // TODO: When is a fork and only a subset of your peers know about a block, sampling should only
+            // be queried on the peers on that fork. Should this case be handled? How to handle it?
```

Comment:
Maybe we could use status messages, similar to how we do it in range_sync - happy to leave this out of this current PR though

https://github.com/sigp/lighthouse/blob/23671531bcb2c8050888b51b80f1be2310384037/beacon_node/network/src/sync/manager.rs#L345-L351

Reply:
It's still tricky because peers can have slightly different heads on the same fork such that they _should_ be grouped together but they will appear in different forks

Reply:
Yep let's think about this - going to resolve this and we'll get back to this TODO later